lace/topology.py

```diff
@@ -1,4 +1,99 @@
 # pylint: disable=attribute-defined-outside-init, access-member-before-definition, len-as-condition
+import zlib
+
+def get_vert_connectivity(mesh):
+    """Returns a sparse matrix (of size #verts x #verts) where each nonzero
+    element indicates a neighborhood relation. For example, if there is a
+    nonzero element in position (15, 12), that means vertex 15 is connected
+    by an edge to vertex 12."""
+    import numpy as np
+    import scipy.sparse as sp
+    from blmath.numerics.matlab import row
+
+    vpv = sp.csc_matrix((len(mesh.v), len(mesh.v)))
+
+    # for each column in the faces...
+    for i in range(3):
+        IS = mesh.f[:, i]
+        JS = mesh.f[:, (i+1)%3]
+        data = np.ones(len(IS))
+        ij = np.vstack((row(IS.flatten()), row(JS.flatten())))
+        mtx = sp.csc_matrix((data, ij), shape=vpv.shape)
+        vpv = vpv + mtx + mtx.T
+
+    return vpv
+
+_vertices_per_edge_cache = {}
+def get_vertices_per_edge(mesh, faces_per_edge=None):
+    """Returns an Ex2 array of adjacencies between vertices, where
+    each element in the array is a vertex index. Each edge is included
+    only once. If output of get_faces_per_edge is provided, this is used to
+    avoid call to get_vert_connectivity()"""
+    import numpy as np
+    import scipy.sparse as sp
+    from blmath.numerics.matlab import row, col
+
+    faces = mesh.f
+    suffix = str(zlib.crc32(faces_per_edge.flatten())) if faces_per_edge != None else ''
+    cache_key = str(zlib.crc32(faces.flatten())) + '_' + suffix
+    if cache_key in _vertices_per_edge_cache:
+        return _vertices_per_edge_cache[cache_key]
+    else:
+        if faces_per_edge != None:
+            #result = np.asarray([vertices_in_common(face_pair[0], face_pair[1]) for face_pair in mesh.f[faces_per_edge]])
+            #new code above is a few times faster
+            result = np.asarray(np.vstack([row(np.intersect1d(mesh.f[k[0]], mesh.f[k[1]])) for k in faces_per_edge]), np.uint32)
+        else:
+            vc = sp.coo_matrix(get_vert_connectivity(mesh))
+            result = np.hstack((col(vc.row), col(vc.col)))
+            result = result[result[:, 0] < result[:, 1]] # for uniqueness
+
+        _vertices_per_edge_cache[cache_key] = result
+        return result
+
+def gen_coupling_weights(sm_template, default_weights=1.0):
```

Comment:
This does not belong in lace. It's scan alignment specific. Please move it back to tfsmpl.